lib/processors/manifestEnhancer.js

```diff
@@ -0,0 +1,417 @@
+	 */
+	constructor(manifest, filePath, fs) {
+		this.fsReadDir = promisify(fs.readdir);
+		this.cwd = path.dirname(filePath);
+		this.manifest = JSON.parse(manifest);
+
+		this.isModified = false;
+		this.runInvoked = false;
+	}
+
+	markModified() {
+		this.isModified = true;
+	}
+
+	async readdir(relativePath) {
+		const absolutePath = path.resolve(this.cwd, relativePath);
+		try {
+			return await this.fsReadDir(absolutePath);
+		} catch (err) {
+			if (err?.code === "ENOENT") {
+				return [];
+			} else {
+				throw err;
+			}
+		}
+	}
+
+	async findSupportedLocales(i18nBundleUrl) {
+		const i18nBundleName = path.basename(i18nBundleUrl, ".properties");
+		const i18nBundlePrefix = `${i18nBundleName}_`;
+		const i18nBundleDir = path.dirname(i18nBundleUrl);
+		const i18nBundleFiles = await this.readdir(i18nBundleDir);
+		const supportedLocales = [];
+		i18nBundleFiles.forEach((fileName) => {
+			if (!fileName.endsWith(".properties")) {
+				return;
+			}
+			const fileNameWithoutExtension = path.basename(fileName, ".properties");
+			if (fileNameWithoutExtension === i18nBundleName) {
+				supportedLocales.push("");
+			} else if (fileNameWithoutExtension.startsWith(i18nBundlePrefix)) {
+				const locale = fileNameWithoutExtension.replace(i18nBundlePrefix, "");
+				supportedLocales.push(locale);
+			}
+		});
+		return supportedLocales.sort();
+	}
+
+	async processBundleConfig({bundleConfig, fallbackBundleUrl, isTerminologyBundle = false, fallbackLocale}) {
+		const bundleUrl = getBundleUrlFromConfigObject(bundleConfig, this.manifest["sap.app"].id, fallbackBundleUrl);
+		if (!bundleUrl) {
+			return;
+		}
+		if (bundleConfig.supportedLocales) {
+			return;
+		}
+
+		const supportedLocales = await this.getSupportedLocales(
+			bundleUrl, fallbackLocale ?? bundleConfig.fallbackLocale, isTerminologyBundle
+		);
+		if (supportedLocales.length > 0) {
+			bundleConfig.supportedLocales = supportedLocales;
+			this.markModified();
+		}
+	}
+
+	async getSupportedLocales(bundleUrl, fallbackLocale, isTerminologyBundle = false) {
+		// Ignore absolute URLs
+		if (isAbsoluteUrl(bundleUrl)) {
+			return [];
+		}
+		const resolvedBundleUrl = resolveUI5Url(bundleUrl);
+		if (!resolvedBundleUrl) {
+			// In case of a relative ui5-protocol URL
+			return [];
+		}
+		const sapAppId = this.manifest["sap.app"].id;
+		const normalizedBundleUrl = normalizeBundleUrl(resolvedBundleUrl, sapAppId);
+		if (normalizedBundleUrl.startsWith("../")) {
+			log.verbose(
+				`bundleUrl '${bundleUrl}' points to a bundle outside of the ` +
+				`current namespace '${sapAppId}', enhancement of 'supportedLocales' is skipped`
+			);
+			return [];
+		}
+		const supportedLocales = await this.findSupportedLocales(normalizedBundleUrl);
+		if (!isTerminologyBundle && supportedLocales.length > 0) {
+			if (fallbackLocale && !supportedLocales.includes(fallbackLocale)) {
+				log.error(
+					`manifest.json: ` +
```

Comment:
I think we should include the virtual path of the affected manifest.json to make it easier to identify which file we are complaining about.

```suggestion
					`${this.filePath}: ` +
```